File: flutter-sdk/lib/src/data/services/yalo_media/yalo_media_service_remote.dart
// Copyright (c) Yalochat, Inc. All rights reserved.

import 'dart:convert';
import 'dart:typed_data';

import 'package:chat_flutter_sdk/src/common/result.dart';
import 'package:chat_flutter_sdk/src/data/services/yalo_media/media_upload_response.dart';
import 'package:chat_flutter_sdk/src/data/services/yalo_media/yalo_media_service.dart';
import 'package:chat_flutter_sdk/src/data/services/yalo_message_auth/token_entry.dart';
import 'package:chat_flutter_sdk/src/data/services/yalo_message_auth/yalo_message_auth_service.dart';
import 'package:http/http.dart';
import 'package:image_picker/image_picker.dart';
import 'package:logging/logging.dart';
import 'package:mime/mime.dart';

class YaloMediaServiceRemote implements YaloMediaService {
  final String _baseUrl;
  final YaloMessageAuthService _authService;
  final Client _httpClient;
  final Logger log = Logger('YaloMediaServiceRemote');

  YaloMediaServiceRemote({
    required String baseUrl,
    required YaloMessageAuthService authService,
    Client? httpClient,
  }) : _baseUrl = baseUrl,
       _authService = authService,
       _httpClient = httpClient ?? Client();

  @override
  Future<Result<MediaUploadResponse>> uploadMedia(XFile file) async {
    final authResult = await _authService.auth();
    if (authResult case Error(:final error)) {
      return Result.error(error);
    }
    final entry = (authResult as Ok<TokenEntry>).result;

    try {
      final bytes = await file.readAsBytes();
      final mimeType = lookupMimeType(file.name) ?? 'application/octet-stream';

      final request = MultipartRequest(
        'POST',
        Uri.parse('$_baseUrl/all/media'),
      );

      log.info('Sending mime type $mimeType');
      request.headers['Authorization'] = 'Bearer ${entry.accessToken}';
      request.files.add(
        MultipartFile.fromBytes(
          'file',
          bytes,
          filename: file.name,
          contentType: MediaType.parse(mimeType),
        ),
      );

      final streamedResponse = await _httpClient.send(request);
      final response = await Response.fromStream(streamedResponse);

      if (response.statusCode == 201) {
        final json = jsonDecode(response.body) as Map<String, dynamic>;
        return Result.ok(MediaUploadResponse.fromJson(json));
      } else {
        log.severe('Unable to upload media', response.body);
        return Result.error(
          Exception('Failed to upload media: ${response.statusCode}'),
        );
      }
    } on Exception catch (e) {
      return Result.error(e);
    }
  }

  @override
  Future<Result<Uint8List>> downloadMedia(String url) async {
    try {
      final response = await _httpClient.get(Uri.parse(url));
      if (response.statusCode == 200) return Result.ok(response.bodyBytes);
      return Result.error(
        Exception('Failed to download media: ${response.statusCode}'),
      );
    } on Exception catch (e) {
      return Result.error(e);
    }
  }
}

File: web-sdk/src/data/services/yalo-media/yalo-media-service-remote.ts
// Copyright (c) Yalochat, Inc. All rights reserved.

import { Err, Ok, type Result } from '@domain/common/result';
import type { TokenRepository } from '@data/repositories/token/token-repository';
import type {
  MediaUploadResponse,
  YaloMediaService,
} from './yalo-media-service';

export class YaloMediaServiceRemote implements YaloMediaService {
  private readonly _baseUrl: string;
  private readonly _tokenRepository: TokenRepository;

  constructor(baseUrl: string, tokenRepository: TokenRepository) {
    this._baseUrl = baseUrl;
    this._tokenRepository = tokenRepository;
  }

  async uploadMedia(file: File): Promise<Result<MediaUploadResponse>> {
    const tokenResult = await this._tokenRepository.getToken();
    if (!tokenResult.ok) return tokenResult;

    try {
      const formData = new FormData();
      formData.append('file', file, file.name);

      const response = await fetch(`${this._baseUrl}/all/media`, {
        method: 'POST',
        headers: {
          Authorization: `Bearer ${tokenResult.value}`,
        },
        body: formData,
      });

      if (response.status === 201) {
        const json = (await response.json()) as MediaUploadResponse;
        return new Ok(json);
      }

      return new Err(new Error(`Failed to upload media: ${response.status}`));
    } catch (e) {
      return new Err(e instanceof Error ? e : new Error(String(e)));
    }
  }

  async downloadMedia(url: string): Promise<Result<Blob>> {
    try {
      const response = await fetch(url);

      if (response.ok) {
        return new Ok(await response.blob());
      }

      return new Err(new Error(`Failed to download media: ${response.status}`));
    } catch (e) {
      return new Err(e instanceof Error ? e : new Error(String(e)));
    }
  }
}
